### ISSUE TITLE
reading gzip compressed files via CodecZlib/TranscodingStreams

### ISSUE THREAD

**Issue by tpapp** [JuliaData/CSV.jl#366]
This was discussed in #46, but now Libz.jl has been replaced mostly by the TranscodingStreams family of libraries, eg CodecZlib, and the current CSV does not work with them.

```julia
julia> io = GzipDecompressorStream(gzio)
TranscodingStreams.TranscodingStream{GzipDecompressor,IOStream}(<mode=idle>)

julia> csv = CSV.File(io)
ERROR: MethodError: no method matching position(::TranscodingStreams.TranscodingStream{GzipDecompressor,IOStream})
Closest candidates are:
  position(::JLD2.MmapIO) at /home/tamas/.julia/packages/JLD2/KjBIK/src/mmapio.jl:253
  position(::JLD2.BufferedWriter) at /home/tamas/.julia/packages/JLD2/KjBIK/src/bufferedio.jl:63
  position(::JLD2.BufferedReader) at /home/tamas/.julia/packages/JLD2/KjBIK/src/bufferedio.jl:113
```
Is there a chance they could be supported? Or is `position` absolutely necessary for this library?

**Comment by bicycle1885** [comment]
I've implemented a `position` method in the latest TranscodingStreams.jl but CSV.jl (to say more accurately, Parsers.jl) requires more methods such as `seek`. I think `TranscodingStream` cannot support `seek` in general because we cannot transcode data backward. I think the best approach is making Parsers.jl depend on TranscodingStreams.jl so that Parsers.jl can use data buffered in `TranscodingStream`.

**Comment by bicycle1885** [comment]
I've quickly read the code of CSV.jl and found that it gathers starting positions of rows in the first pass and then loads the records, which requires the `seek` function. So, I think TranscodingStreams.jl and CSV.jl cannot work together...

**Comment by nalimilan** [comment]
Could TranscodingStreams at least support `seek(s, 0)` to restart reading from scratch? In the most common pattern I guess CSV only needs that.

**Comment by bicycle1885** [comment]
Yes, it is possible if the wrapped stream also supports the same operation, though I need to slightly modify the code of TranscodingStreams.jl. However, it still requires decompressing the same data twice. I'm interested in the reason why CSV.jl takes such a little bit strange strategy.

**Comment by nalimilan** [comment]
That's because CSV is a terrible format which doesn't carry any information about column types, so we need to parse a sample of rows to detect them. If you provide types manually a single pass is needed AFAIK.

**Comment by bicycle1885** [comment]
Yeah, but CSV.jl seems to read the whole stream before it starts parsing by default: https://github.com/JuliaData/CSV.jl/blob/147ba7a86cc556743d91d62850b3f5212f6246ca/src/filedetection.jl#L215-L248. This requires decompressing, going back to the beginning, and again decompressing. I think this is not a cheap operation especially when the stream is compressed with some format such as xz.

**Comment by nalimilan** [comment]
Ah, that's to count the number of rows. Not sure this can be avoided: the number of rows is an interesting information to preallocate the storage.

**Comment by tpapp** [comment]
@bicycle1885, thanks for doing this. Unfortunately it seems that the design of CSV.jl and Parsers.jl is based on the assumption that the contents of the CSV file are readily available and can be traversed in an arbitrary way at little additional cost. This allows various optimizations that make reading really fast, but at the same time rule out working from streams.

This is a reasonable choice in some contexts, but it would be great if the package README and documentation **made this explicit, so that users could make an informed choice**. Especially since many people have a completely different use case: they have very large CSV files, which are very compressible, and need to be ingested and converted to another format rarely enough so that economizing on disk space is preferred.

I am not asking that CSV.jl supports this use case if it is at odds with the design of the package; just that this is clarified. Then, as far as I am concerned, this issue could be closed.

**Comment by nalimilan** [comment]
I don't think this assumption is only used for optimizations. You basically can't determine the column types without reading a sample of rows first. Do you know of other CSV parsers which don't do that?

Also it seems CSV.jl with its current approach can be made to work with compressed files. It will just trigger decompression several times, which isn't great but certainly better than nothing. In particular, it shouldn't require you to store the full decompressed file.

**Comment by tpapp** [comment]
> You basically can't determine the column types without reading a sample of rows first.

This is outside the scope of this issue, but a package could just try parsing each cell as `Int`, `Float64`, and `String` (in this order, the last one is valid for everything, and none of these lose information), and `push!` these into one of 3 `Vector`s, perhaps making allowances for `missing` etc, and when the file is done make a heuristic decision about converting to a common column type or not.

I respect the design choices of CSV.jl, but I don't think they are the only reasonable one.

**Comment by nalimilan** [comment]
Actually it's more complex than that: if you parse a column as integer and it succeeds, but then you encounter a value which isn't numeric, it is too late to recover the data you have already parsed as integers. For example, `"0001"` will have been converted to `1`, which when converted to strings will give `"1"`. So basically the only solutions are: 1) know the types in advance, 2) detect them in advance and assume they won't change (this package), 3) keep a copy of all entries in a `Vector{String}` until parsing is done. 3. is mostly equivalent to storing a decompressed copy of the file in memory. At that point you could as well use a non-streaming decompression package.

It would be interesting to know whether some CSV parsers take a different approach. I suspect there can be variants to avoid decompressing everything (e.g. do not count rows beforehand, only use a few rows from the beginning of the file for type detection), but the core design is probably quite constant.

**Comment by tpapp** [comment]
Again, this is somewhat orthogonal to this issue, but since CSV is a very loosely specified format, and does not strictly distinguish strings that look like numbers from numbers, and neither does it impose that columns have to have the same element type, either keeping `1::Int` in a `Vector{Union{Int,String}}` or converting back to `"1"` are valid choices.

In practice all nontrivial CSV parsing requires some post-processing, and frequently preprocessing. I recently parsed some TSV data from EuroStat, which had fields like `"5.5 p"` where `p` denoted "predicted", and about 1% of fields were like this, the rest were normal numbers like `"5.5"`.

There is no such thing as canonical CSV parsing, only various heuristics. Again, I respect the choices that CSV.jl made, since they are reasonable, but it is hard to argue that they are the "right" ones in any sense.

**Comment by quinnj** [comment]
This will be fixed in #395

**Comment by quinnj** [comment]
Fixed in #395
